pkg/meta/base.go

```diff
@@ -1438,6 +1442,66 @@ func (m *baseMeta) CleanupTrashBefore(ctx Context, edge time.Time, increProgress
 	}
 }
 
+func (m *baseMeta) scanTrashFiles(ctx Context, scan trashFileScan) error {
+	var st syscall.Errno
+	var entries []*Entry
+	if st = m.en.doReaddir(ctx, TrashInode, 1, &entries, -1); st != 0 {
+		return errors.Wrap(st, "read trash")
+	}
+
+	type scannedEntry struct {
+		deletedAt time.Time
+		*Entry
+	}
+
+	scannedEntries := make([]scannedEntry, 0, len(entries))
+	for _, e := range entries {
+		ts, err := time.Parse("2006-01-02-15", string(e.Name))
+		if err != nil {
+			logger.Warnf("bad entry as a subTrash: %s", e.Name)
+			continue
+		}
+		scannedEntries = append(scannedEntries, scannedEntry{
+			deletedAt: ts,
+			Entry:     e,
+		})
+	}
+
+	for len(scannedEntries) != 0 {
+		poped := scannedEntries
+		scannedEntries = nil
+		for _, entry := range poped {
+			if entry.Attr.Typ == TypeFile {
+				clean, err := scan(entry.Inode, entry.Attr.Length, entry.deletedAt)
+				if err != nil {
+					return errors.Wrap(err, "scan trash files")
+				}
+				if clean {
+					st = m.en.doUnlink(ctx, entry.Attr.Parent, string(entry.Name))
+					if st != 0 {
+						return errors.Wrap(st, "unlink trash file")
+					}
+				}
+				continue
+			}
+			if entry.Attr.Typ == TypeDirectory {
+				var subEntries []*Entry
+				if st = m.en.doReaddir(ctx, entry.Inode, 1, &subEntries, -1); st != 0 {
```

Comment:
The directory should be empty

Reply:
Why should the directory be empty? Here we won't remove the directory, just read it and push subentries into the queue.

Reply:
The trash only has three levels,  trash/directory/file or empty directory. 
We should avoid the `readdir` for empty directory.

Reply:
Fixed, the new implementation ignores empty directories.